File: 36 Valid Sudoku/Java/Solution.java
public class Solution {
    public boolean isValidSudoku(char[][] board) {
        for (int r = 0; r < board.length; r++) {
            for (int c = 0; c < board[0].length; c++) {
                if (board[r][c] != '.' && !isValidPlacement(board, board[r][c], r, c)) {
                    return false;
                }
            }
        }

        return true;
    }


    public static boolean isValidPlacement(char[][] board, char num, int row, int col) {
        return !isNumInRow(board, num, row, col) && !isNumInCol(board, num, row, col) && !isNumInBox(board, num, row, col);
    }


    public static boolean isNumInRow(char[][] board, char num, int row, int col) {
        for (int c = 0; c < board[0].length; c++) {
            if (c != col && board[row][c] == num) {
                return true;
            }
        }

        return false;
    }


    public static boolean isNumInCol(char[][] board, char num, int row, int col) {
        for (int r = 0; r < board.length; r++) {
            if (r != row && board[r][col] == num) {
                return true;
            }
        }

        return false;
    }


    public static boolean isNumInBox(char[][] board, char num, int row, int col) {
        int boxRow = row - row % 3;
        int boxCol = col - col % 3;

        for (int r = boxRow; r < boxRow+3; r++) {
            for (int c = boxCol; c < boxCol+3; c++) {
                if (r != row && c != col && board[r][c] == num) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: 36 Valid Sudoku/Python/Solution.py
from typing import List, Self


class Solution:
    def isValidInRow(self: Self, board: List[List[str]], row: int, col: int) -> bool:
        row: List[str] = board[row]
        for c in range(len(row)):
            if c != col and row[col] == row[c]:
                return False
        return True
    
    def isValidInCol(self: Self, board: List[List[str]], row: int, col: int) -> bool:
        for r in range(len(board)):
            if r != row and board[row][col] == board[r][col]:
                return False
        return True
    
    def isValidInGrid(self: Self, board: List[List[str]], row: int, col: int) -> bool:
        rStart: int = row // 3 * 3
        cStart: int = col // 3 * 3
        for deltaR in range(3):
            for deltaC in range(3):
                r: int = rStart + deltaR
                c: int = cStart + deltaC
                if r != row and c != col and board[r][c] == board[row][col]:
                    return False
        return True

    def isValid(self: Self, board: List[List[str]], row: int, col: int) -> bool:
        return self.isValidInRow(board, row, col) and self.isValidInCol(board, row, col) and self.isValidInGrid(board, row, col)
    
    def isValidSudoku(self: Self, board: List[List[str]]) -> bool:
        for r in range(len(board)):
            for c in range(len(board[r])):
                if board[r][c] != '.' and not self.isValid(board, r, c):
                    return False
        return True